### ISSUE TITLE
Up-to-date check should not look at imported targets/props

### ISSUE THREAD

**Issue by panopticoncentral** [dotnet/project-system#2415]
When implementing the new up-to-date check, we copied the behavior from C++ of checking timestamps of imported targets/props files. While this is good in that it will cause projects to rebuild if someone changes a targets/props file that the project depends on, it's bad in that if those changes don't actually cause the outputs to change, the project will not be considered up-to-date again until it is caused to be rebuilt by a code change or a Rebuild or something like that.

The simplest solution is to stop looking at imported targets/props, which is I believe what the old up-to-date check did. Another solution would be to do something like try to force the outputs to be rebuilt somehow, but that seems pretty intrusive.

What do you think, @srivatsn @davkean ?

**Comment by sharwell** [comment]
You could also use an intermediate file which just writes the names of all the imported targets files, and make sure that file is up-to-date with respect to the props/targets files.

**Comment by davkean** [comment]
Not sure I understand why this behavior isn't expected, can you expand on this?

**Comment by davkean** [comment]
I think it would be odd, also that the design-time build would pick up the "changes" - but the full build wouldn't.

**Comment by panopticoncentral** [comment]
The bad-behavior scenario would be this:

- User is working in Roslyn.sln
- User builds Roslyn.sln, all projects are now considered up to date
- User can work in single project, all others are considered up to date
- User makes a change to one of the Roslyn targets files, maybe something related to deployment
- User builds Roslyn.sln. Since every project's output is older than one of the project's imports, none are considered up to date. But since the targets change didn't affect output, no projects are actually built.
- From now on, all projects that haven't been touched will be considered out of date because their outputs will be older than one of the project's imports

Basically, all projects will continue to be out of date until the user does a Rebuild.

My understanding is that MSBuild actually doesn't check project imports timestamps to determine whether a project file is up to date, so we're actually being more strict here than MSBuild.

If we think this is a problem, then I think we should address it first at the MSBuild level along the lines of what @sharwell proposes, and then the up to date check can look at that. This is similar to what we're doing with the ref assemblies up to date check.

**Comment by davkean** [comment]
@panopticoncentral See the `<MSBuildAllProjects></MSBuildAllProjects>` property, that is equivalent of MSBuild checking the targets. Perhaps we should be respecting that? It is an input for CoreCompile: https://github.com/dotnet/roslyn/blob/2ea8ebbc5ac5f860d690c6b2c05be19e0812461b/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L17.

Though I can still see an issue with determinism turned on, CoreCompile might not produce an updated output, even in that case.

**Comment by davkean** [comment]
@rainersigwald What's your thoughts on this?

**Comment by rainersigwald** [comment]
I think the imports should invalidate FUTD. You could, for example, tweak a constant definition in a props file that would affect projects that aren't having other inputs changed (but can observe differences in behavior from the different definitions on references). I'd also find it quite surprising if I tweaked something in an imported file and hit F5 and _nothing_ built, not even the project I was "logically" working on.

> Though I can still see an issue with determinism turned on, CoreCompile might not produce an updated output.

I don't think that's a problem at the moment, because the compiler always emits an output with a new timestamp -- it just might be bitwise identical to the previous one. That is then copied around as it always has been. (@jcouv to confirm -- should we consider using `CopyRefAssembly` to copy the implementation assembly around as well as the ref assembly?)

**Comment by rainersigwald** [comment]
Reading more, I think the key is that `$(MSBuildAllProjects)` is an input for `CoreCompile` -- so this

> But since the targets change didn't affect output, no projects are actually built.

shouldn't be true for standard-targets projects.

MSBuild doesn't consider import timestamps at the engine level, but "wise" targets often do using `$(MSBuildAllProjects)` (which shouldn't have to be manually maintained but currently is: https://github.com/Microsoft/msbuild/issues/1299).

**Comment by davkean** [comment]
Basically, then it comes down to the determination of whether we look at all targets or just those in $(MSBuildAllProjects). $(MSBuildAllProjects) would be consistent with MSBuild itself until Microsoft/msbuild#1299 is implemented.

**Comment by panopticoncentral** [comment]
Right, this issue came about because I was testing something and had caused some target file to update, and my project was always failing the up to date check and calling MSBuild. Unfortunately, I didn't actually record what target file it was--it sounds like it was probably a bad target file that didn't add itself to MSBuildAllProjects, because when MSBuild was being called it wasn't rebuilding the output of the project.

If we believe that we're going to fix Microsoft/msbuild#1299, then I think we should leave the current behavior as-is because it matches what is going to happen and it's simpler that parsing out MSBuildAllProjects. Changing targets should be a fairly exotic activity and the situation can be easily fixed by doing a Rebuild.

Put another way, the project system is correct and MSBuild has a bug/design flaw, which has been logged and will be fixed. If that sounds good to everyone, I'll close this issue.

**Comment by rainersigwald** [comment]
I'd like to fix Microsoft/msbuild#1299, but it'll be no earlier than 15.6 and more likely 16.0. As long as everybody's ok with that timeline, I'm ok with closing this.

**Comment by davkean** [comment]
I'm okay with that. For posterity sake, the workaround for those affected by this would be to add the following to the target that is affected:

``` XML
  <PropertyGroup>
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>
```
